aws/src/main/java/org/apache/iceberg/aws/s3/signer/S3V4RestSignerClient.java

```diff
@@ -0,0 +1,311 @@
+    } else {
+      Map<String, String> responseHeaders = Maps.newHashMap();
+      Consumer<Map<String, String>> responseHeadersConsumer = responseHeaders::putAll;
+      S3SignResponse s3SignResponse =
+          httpClient()
+              .post(
+                  endpoint(),
+                  remoteSigningRequest,
+                  S3SignResponse.class,
+                  () -> authSession().headers(),
+                  ErrorHandlers.defaultErrorHandler(),
+                  responseHeadersConsumer);
+
+      signedComponent =
+          ImmutableSignedComponent.builder()
+              .headers(s3SignResponse.headers())
+              .signedURI(s3SignResponse.uri())
+              .build();
+
+      if (canBeCached(responseHeaders)) {
+        SIGNED_COMPONENT_CACHE.put(cacheKey, signedComponent);
+      }
+    }
+
+    // The SdkHttpFullRequest Builder appends the raw path from the input URI in .uri(),
+    // so we need to clear the current path from the request
+    SdkHttpFullRequest.Builder mutableRequest = request.toBuilder();
+    mutableRequest.encodedPath("");
+    mutableRequest.uri(signedComponent.signedURI());
+    reconstructHeaders(signedComponent.headers(), mutableRequest);
+
+    return mutableRequest.build();
+  }
+
+  private void reconstructHeaders(
+      Map<String, List<String>> signedAndUnsignedHeaders,
+      SdkHttpFullRequest.Builder mutableRequest) {
+    Map<String, List<String>> headers = Maps.newHashMap(signedAndUnsignedHeaders);
+    // we need to remove the Cache-Control header that is being sent by the server
+    headers.remove(CACHE_CONTROL);
+
+    // we need to overwrite whatever headers the server signed/unsigned with the ones from the
+    // original request and then put all headers back to the request
+    headers.putAll(mutableRequest.headers());
+    headers.forEach(mutableRequest::putHeader);
+  }
+
+  private boolean canBeCached(Map<String, String> responseHeaders) {
+    return responseHeaders.containsKey(CACHE_CONTROL)
+        && CACHE_CONTROL_PRIVATE.equals(responseHeaders.get(CACHE_CONTROL));
+  }
+
+  private void checkSignerParams(AwsS3V4SignerParams signerParams) {
+    if (signerParams.enablePayloadSigning()) {
+      throw new UnsupportedOperationException("Payload signing not supported");
+    }
+
+    if (signerParams.enableChunkedEncoding()) {
+      throw new UnsupportedOperationException("Chunked encoding not supported");
+    }
+  }
+
+  @Value.Immutable
+  interface Key {
+    String method();
+
+    String region();
+
+    String uri();
+
+    static Key from(S3SignRequest request) {
+      return ImmutableKey.builder()
+          .method(request.method())
+          .region(request.region())
+          .uri(request.uri().toString())
+          .build();
+    }
+  }
+
+  @Value.Immutable
+  interface SignedComponent {
+    Map<String, List<String>> headers();
+
+    URI signedURI();
+  }
+
+  public static S3V4RestSignerClient of(Map<String, String> properties) {
```

Comment:
nit: might want to add a javadoc here describing this class is used for dynamic initialization

Reply:
just fyi, we discussed with @danielcweeks offline and decided to rename it to `create` to cause less confusion. I don't think the javadoc is required, since it's not "only" for dynamic loading. It's just rather a quicker way to init an instance than going through the Immutable builder